design/0001-frr.md

```diff
@@ -0,0 +1,341 @@
+MetalLB’s current built-in BGP speaker.
+
+## Motivation
+
+The motivation for this enhancement is to provide an alternative
+production-ready BGP agent for use in MetalLB. Overall, this should reduce the
+effort for adding additional features to the MetalLB project. For example, there
+are a number of [issues](https://github.com/metallb/metallb/issues?q=is%3Aissue+is%3Aopen+label%3Aprotocol%2FBGP)
+in the current backlog that may be addressed by using FRR. Notably:
+
+* [Add support for MP BGP encoding for IPv4 and IPv6](https://github.com/metallb/metallb/pull/590)
+* [BFD support](https://github.com/metallb/metallb/issues/396)
+* [BGP Failover too slow](https://github.com/metallb/metallb/issues/485)
+* [OSPF Support](https://github.com/metallb/metallb/issues/256)
+* [RIP Support](https://github.com/metallb/metallb/issues/9)
+* [Add IPv6 BGP support](https://github.com/metallb/metallb/issues/7)
+
+FRR is a mature Linux Foundation routing protocol suite based on Quagga that has
+been used in many production deployments. As such, it has been proven in terms
+of its maturity, flexibility (as can be seen by the [broad range of features](http://docs.frrouting.org/en/latest/bgp.html)
+it supports), scalability, security, reliability and performance. It also
+provides detailed logging features to aid debugging.
+
+The current native BGP implementation of MetalLB has limitations when coexisting
+with other BGP agents on the node. If these BGP agents are attempting to peer
+with external routers, this may invoke BGP’s conflict resolution algorithm as
+only one agent on a node may peer with a remote BGP agent. This is detailed
+[here](https://metallb.universe.tf/configuration/calico/) for the Calico case
+and [here](https://metallb.universe.tf/configuration/kube-router/) for the
+Kube-Router case. This enhancement may be a step towards resolving these
+issues by potentially allowing MetalLB to reuse a BGP agent already in use on
+the node. This local agent could then establish the session with the remote
+agent and be the single entity peering with the remote agent. In this way,
+MetalLB and other software on the node could share the BGP session with the
+remote agent. For the Calico case, this would require integrating BIRD as an
+alternative BGP implementation as Calico uses BIRD for its BGP functionality.
+This would also require additional work to configure FRR in a way that would
+allow it to be shared.
+
+### Goals
+
+* Provide a configuration option to allow use of alternate BGP agents in a
+  MetalLB deployment at cluster installation time.
+* Ensure feature parity with current BGP implementation in order to enable
+  the eventual retirement of the current native BGP implementation.
+* Reinstate thorough BGP testing infrastructure.
+* Add additional documentation to indicate how to troubleshoot and debug any
+  issues with FRR.
+* This will not expose direct configuration of FRR to the MetalLB user.
+
+### Non-Goals
```

Comment:
Ok, I have rephrased this, simplified it and also moved "This will not expose direct configuration of FRR to the MetalLB user." to non-goals as "Expose direct configuration of FRR to the MetalLB user."